File: backend/models/spotify.go
package models

type Track struct {
	Album       Album              `json:"album"`
	Artists     []SimplifiedArtist `json:"artists"`
	DurationMS  int                `json:"duration_ms"`
	Explicit    bool               `json:"explicit"`
	Href        string             `json:"href"`
	ID          string             `json:"id"`
	Name        string             `json:"name"`
	Popularity  int                `json:"popularity"`
	TrackNumber int                `json:"track_number"`
	Type        string             `json:"type"`
	Uri         string             `json:"uri"`
}

//type Playlist struct {
//	ID       string  `json:"id"`
//	Href     string  `json:"href"`
//	Items    []Item  `json:"items"`
//	Limit    int     `json:"limit"`
//	Next     string  `json:"next"`
//	Offset   int     `json:"offset"`
//	Previous string  `json:"previous"`
//	Total    int     `json:"total"`
//	PlaylistTracks   []Track `json:"tracks"`
//}

type PlaylistTracks struct {
	Href     string          `json:"href"`
	Limit    int             `json:"limit"`
	Next     string          `json:"next"`
	Offset   int             `json:"offset"`
	Previous string          `json:"previous"`
	Total    int             `json:"total"`
	Items    []PlaylistTrack `json:"items"`
}

type PlaylistTrack struct {
	AddedAt string `json:"added_at"`
	//AddedBy string `json:"added_by"`
	IsLocal bool  `json:"is_local"`
	Track   Track `json:"track"`
}

type Playlists struct {
	Href     string         `json:"href"`
	Limit    int            `json:"limit"`
	Next     string         `json:"next"`
	Offset   int            `json:"offset"`
	Previous string         `json:"previous"`
	Total    int            `json:"total"`
	Items    []PlaylistItem `json:"items"`
}

type PlaylistItem struct {
	Collaborative bool    `json:"collaborative"`
	Description   string  `json:"description"`
	Href          string  `json:"href"`
	ID            string  `json:"id"`
	Images        []Image `json:"images"`
	Name          string  `json:"name"`
	Public        bool    `json:"public"`
	SnapshotID    string  `json:"snapshot_id"`
	//PlaylistTracks        []Track `json:"tracks"`	// Review
	Type string `json:"type"`
	URI  string `json:"uri"`
}

type TopTracks struct {
	Href     string  `json:"href"`
	Limit    int     `json:"limit"`
	Next     string  `json:"next"`
	Offset   int     `json:"offset"`
	Previous string  `json:"previous"`
	Total    int     `json:"total"`
	Items    []Track `json:"items"`
}

type Album struct {
	AlbumType   string             `json:"album_type"`
	TotalTracks int                `json:"total_tracks"`
	Href        string             `json:"href"`
	Id          string             `json:"id"`
	Images      []Image            `json:"images"`
	Name        string             `json:"name"`
	Type        string             `json:"type"`
	Uri         string             `json:"uri"`
	Artists     []SimplifiedArtist `json:"artists"`
}

type SimplifiedArtist struct {
	Href string `json:"href"`
	Id   string `json:"id"`
	Name string `json:"name"`
	Type string `json:"type"`
	Uri  string `json:"uri"`
}

type Image struct {
	URL    string `json:"url"`
	Height int    `json:"height"`
	Width  int    `json:"width"`
}

type UserProfile struct {
	Country     string  `json:"country"`
	DisplayName string  `json:"display_name"`
	Email       string  `json:"email"`
	Href        string  `json:"href"`
	ID          string  `json:"id"`
	Images      []Image `json:"images"`
	Product     string  `json:"product"`
	Type        string  `json:"type"`
	URI         string  `json:"uri"`
}

type Token struct {
	AccessToken  string `json:"access_token"`
	TokenType    string `json:"token_type"`
	ExpiresIn    int    `json:"expires_in"`
	RefreshToken string `json:"refresh_token"`
	Scope        string `json:"scope"`
}

File: frontend/src/types/Spotify.ts
export interface PlaylistTracks {
  href: string;
  limit: number;
  next: string;
  offset: number;
  previous: string;
  total: number;
  items: PlaylistTrack[];
}

export interface PlaylistTrack {
  added_at: string;
  // added_by: {
  //   id: string;
  //   name: string;
  // };
  is_local: boolean;
  track: Track;
}

export interface Track {
  album: Album;
  artists: SimplifiedArtist[];
  duration_ms: number;
  explicit: boolean;
  href: string;
  id: string;
  name: string;
  popularity: number;
  preview_url: string;
  track_number: number;
  type: string;
  uri: string;
}

export interface Image {
  url: string;
  height: number;
  width: number;
}

export interface PlaylistItem {
  collaborative: boolean;
  description: string;
  href: string;
  id: string;
  images: Image[];
  name: string;
  public: boolean;
  snapshot_id: string;
  // tracks: {
  //   href: string;
  //   total: number;
  // };
  type: string;
  uri: string;
}

export interface Album {
  album_type: string;
  total_tracks: number;
  href: string;
  id: string;
  images: Image[];
  name: string;
  type: string;
  uri: string;
  artists: SimplifiedArtist[];
}

export interface SimplifiedArtist {
  href: string;
  id: string;
  name: string;
  type: string;
  uri: string;
}
